Resolve issue with custom tests names and test Ids

johnmwright opened the pull request:
This PR is to address #622 

Key highlights:

- The FullyQualifiedName is based on Type+Method and independent of DisplayName.
- The TestCaseId is unique across all test cases (including data-driven ones) and deterministic between discovery sessions.

## Background information:
- The VS test system [expects the `TestCase.FullyQualifiedName` to be the actual test method name](https://github.com/microsoft/vstest-docs/blob/master/RFCs/0017-Managed-TestCase-Properties.md#specification) (not the display name), optionally including param _types_ (not values). Basically, the details that would have been returned from `caller.GetType().FullName`
- If you don't provide a value for `TestCase.Id`, the class with [create a GUID](https://github.com/microsoft/vstest/blob/829659f9ca95ccfb54480948d8e472651dd0ff5d/src/Microsoft.TestPlatform.ObjectModel/TestCase.cs#L96) value [based on the `TestCase.FullyQualifiedName`](https://github.com/microsoft/vstest/blob/829659f9ca95ccfb54480948d8e472651dd0ff5d/src/Microsoft.TestPlatform.ObjectModel/TestCase.cs#L240). 
- When running tests, if you instruct the system to run a subset of tests (ie: not "all" tests), a filter is provided by VS that includes a list of tests by `FullyQualifiiedName`, which the NUnitAdapter attempt to apply in `TfsTestFilter.CheckFilter(TestCase)`.
- The test hierarchy in the VS Test Explorer UI uses the `FullyQualifiedName` to determine the namespace groupings
- If multiple tests results are associated with the same TestCase instance (ie: `TestCase.Id` is equal), then [it will show the test once in the UI, with multiple result values](https://github.com/microsoft/vstest-docs/blob/master/RFCs/0017-Managed-TestCase-Properties.md#tests-returning-multiple-results), such as this (which is how MSTest-based tests display results when a method is used for test cases):
![image](https://user-images.githubusercontent.com/3373249/67701209-88a12080-f97d-11e9-8f8c-d271a330c377.png)


## How this issue manifested:
- The [current logic](https://github.com/nunit/nunit3-vs-adapter/blob/e12dcb031828181148397058ab43d07e1db2fcad/src/NUnitTestAdapter/TestConverter.cs#L167) sets the `DisplayName` and `FullyQualifiedName` to the display name of the test, which break the contract with VS if the user has utilized the `SetName()` method to override the display name value.  While VS doesn't strictly enforce the `FullyQualifiedName` is an actual method, it does have [some minor validation](https://github.com/microsoft/vstest/blob/829659f9ca95ccfb54480948d8e472651dd0ff5d/src/Microsoft.TestPlatform.Common/Filtering/FilterExpression.cs#L114) to make sure some special characters are used as expected, such as paired parenthesis, resulting in an error such as:
> An exception occurred while invoking executor 'executor://nunit3testexecutor/': Incorrect format for TestCaseFilter Error: Missing '('. Specify the correct format and try again. Note that the incorrect format can lead to no test getting executed.

## How I addressed the issue:
- First, but setting the `TestCase.FullyQualfiedName` to the actual test method name, this conforms with the expectations from VS and fixes the parsing issues.  Since the NUnit-parsed data includes a `classname` and `methodname` attribute, the FQN can be created from that (or, at least, a close-enough version).
- By using the method type+name for the `FullyQualifiedName`, if the adapter doesn't explicitly set the `TestCast.Id`, it will result in each TestCase having the default Id based on the FQN, as described in the Background Information above. Meaning: When the adapter recreates the TestCase object for association with the TestResult object returned with the test results, it would return several results with the same TestCase.Id value, since each of those would have the same FQN. This resulted in the UI displaying the results under the same Test (similar to the screenshot above).  This does not match the existing behavior, where each test case is listed as it's own node in the test case explorer.  Additionally, only the first DisplayName was shown in the UI, since it ignored tests with identical Ids, due to [Ids needing to be unique](https://github.com/microsoft/vstest-docs/blob/master/RFCs/0017-Managed-TestCase-Properties.md#uniqueness). Since NUnit already applies a unique id to each test, which is consistent across test runs, I decided to resolve this issue by assigning the `TestCase.Id` property using the NUnit id attribute value.  This will give each TestCase a unique Id even if they have the same `FullyQualifiedName`, resulting in the individual nodes that is currently used.


## Impacts of changes
I primarily used two sets of tests while working on this PR.  The first is a test class I put together from the use cases listed in #622, which I put into [this gist](https://gist.github.com/johnmwright/bf374e3ac588fd9ceb8476148578772c) and ultimately added as a unit test in this PR.  The other is the [sample solution](https://github.com/johnjaylward/nUnitExpressionTest) provided in comments to #622 . I'll be using these to describe the impacts of this change.

Before the changes, when the problematic parameterized tests were run, you would see:
- the actual name of the test (example: `Repo_jwright.DemoTest2` in screenshot 1), in addition to the parameterized custom names (ex: `Repo_jwright.Name with mismatched quote '"c'`) as peers.
- Tests which had mismatched special characters were represented as inconclusive (aka "Not Run") at the end of the test run, due to the "Incorrect format for TestCaseFilter" error (ex: `Repo_jwright.DemoTest2` in screenshot 1)
- Parameterized tests were grouped across many nodes when custom names were given (see various nodes for `ExpressionTransatorTest` in screenshot 2)


Screenshot 1:
![image](https://user-images.githubusercontent.com/3373249/67712575-40402d80-f992-11e9-8d64-b701747056db.png)

Screenshot 2:
![image](https://user-images.githubusercontent.com/3373249/67712676-72518f80-f992-11e9-9f03-581a5edc66ba.png)


After the changes:
- The Test name itself is no longer a node with results, but rather it's a grouping node (see: `Repo_jwright.DemoTest2` in screenshot 1)
- Tests which have mismatched special characters no longer throw an error and now show their proper results
- Parameterized tests are grouped under the same node based on the FullyQualifiedTestName value (see screenshot 2)


Screenshot 1:
![image](https://user-images.githubusercontent.com/3373249/67712749-92814e80-f992-11e9-8016-966503a4cf45.png)

![image](https://user-images.githubusercontent.com/3373249/67712967-f1df5e80-f992-11e9-8207-9960159ae93d.png)


## Possible Breaking Change
Because this PR changes how the `TestCase.Id` property is generated, this will result in test cases having different IDs than they did before the change was implemented.  Any systems that expect the test ID value to remain the same across history will lose their tracking of test cases.  

This is specifically called out in the inline comments of [`EqtHash.GuidFromString`](https://github.com/microsoft/vstest/blob/829659f9ca95ccfb54480948d8e472651dd0ff5d/src/Microsoft.TestPlatform.ObjectModel/Utilities/EqtHash.cs#L26) in the VSTest codebase, but with the references between TFS/MsftDevOps Bugs and TFS/MsftDevOps workitems.  I am unsure what, if any, associations TFS/MsftDevOps places with test run results.

johnmwright commented:
Note: I believe this change may fix other open issues. I will associate them to this PR as I verify them.

johnmwright commented:
Note: I see the CI failures and will investigate them as soon as I can.

johnmwright commented:
This "Issue" (placeholder) is related to this PR: https://github.com/nunit/nunit3-vs-adapter/issues/505

johnjaylward commented:
@johnmwright This looks like a huge improvement. I love that all my tests now show under a single grouping as that's how I expected them to show in the first place. Great work, and a wonderful explanation in the PR!

It looks like some test cases may need to be updated though to match the changes: 
```
Expected: "NUnit.Tests.FixtureWithTestCases.MethodWithParameters(9,11)"
  But was:  "NUnit.Tests.FixtureWithTestCases.MethodWithParameters"
```

johnmwright commented:
Yeah, there's a few test failures that I missed in my local runs.  I'll need some time to address them.

aitch-hat commented on src/NUnitTestAdapter/TestConverter.cs:
```diff
@@ -159,20 +160,33 @@ public struct TestResultSet
+            // However, the "fullname" from NUnit's file generation is the custom name of the test, so
+            // this code must convert from one to the other.
+            // Reference: https://github.com/microsoft/vstest-docs/blob/master/RFCs/0017-Managed-TestCase-Properties.md
+            var fullname = className + "." + methodName;
```
Nit: It might be worth using a variable name other than `fullname` so that this isn't confused with `testNode.GetAttribute("fullname")`.

aitch-hat commented on src/NUnit.TestAdapter.Tests.Acceptance/TestSourceWithCustomNames.cs:
```diff
@@ -0,0 +1,130 @@
+                        }
+
+                        [Test]
+                        [TestCase(typeof(IEnumerable<(String oneValue, Int32 twoValue)>))]
```
What is the reason for using `String` and `Int32` here when `string` and `int` have been used in the rest of the class?

johnmwright replied:
Oh, that was unintended and is an artifact of me copy&pasting it from another tool while I worked on it.  I'll update them in my next commit.

johnmwright commented:
I've updated the logic (and my description in "How I addressed the issue:") to be more accurate and handle more cases that were failing to parse the passed-in filters from VS (such as `ValueTuples` in the TestCase parameter signature, which were not being properly passed-in by VS).

johnmwright commented:
@johnjaylward Thanks for this! It's been fixed in my latest push.

> 
> It looks like some test cases may need to be updated though to match the changes:
> 
> ```
> Expected: "NUnit.Tests.FixtureWithTestCases.MethodWithParameters(9,11)"
>   But was:  "NUnit.Tests.FixtureWithTestCases.MethodWithParameters"
> ```

johnmwright commented on .gitignore:
```diff
@@ -329,4 +329,6 @@ tools/
-.history
\ No newline at end of file
+.history
+/Dump
```
Ignoring the "dump" files created when troubleshooting builds using the [`DumpXml` utility](https://github.com/nunit/nunit3-vs-adapter/blob/bddee2c7739f9cc91a021aadf9b3a4f2ddba4c2c/src/NUnitTestAdapter/DumpXml.cs#L42).

johnmwright commented:
@peterwald thanks!  I realized I forgot to put a TL;DR at the top of the PR description, so I borrowed your "key highlights" for that purpose :)

OsirisTerje commented on .gitignore:
```diff
@@ -329,4 +329,6 @@ tools/
-.history
\ No newline at end of file
+.history
+/Dump
```
Not sure why you need this one ?  The dump files [normally] goes below the bin folders, which is ignored.   No harm though.....

johnmwright replied:
@OsirisTerje It creates one name `D_FileThatDoesntExist.dll.dump` in the top-level folder.  Not sure which test specifically created it, but it showed up in my git changes :)

OsirisTerje commented:
@johnmwright  Thank you for this awesome work!   I believe this PR is the best one we have ever received.  It is straight to the point, and the documentation you have added is way above anything else I have seen!  

Not only is the documentation comprehensive and well structured,  the code simple and readable, tests have been added, but you have also taken the time to verify the resolution to a bunch of other similar issues!    

Really appreciate this work!
